ls1intum/Artemis
# [Feature] Grading System create/edit page

### Checklist

### Motivation and Context


The instructors wishing to use the grading system in their courses/exams need to have a page where they could create and edit it.

### Description

I've created the grading system create/edit page and made it accessible via a 'Grading System' button on the course-detail and exam-detail pages.

### Steps for Testing


1. Log in to Artemis as an Instructor
2. Navigate to Course Administration
3. Navigate to
  a) Course detail page and check if the Grading Key button is present
  b) Exam detail page and check if the Grading Key button is present
  c) Follow the button to the grading system page
4. Check if the 'help' and 'generate default grading scale' buttons are working properly
5. Try saving both valid and invalid grading scales and check the alerts at the top of the page
  a) valid would be - all percentages from 0 to 100 are covered by grade steps (doubles for percentage values should work fine), adjacent grade steps have overlapping percentages at the boundaries, all grade steps have nonempty names/bonus points
  b) invalid would be - leaving empty name fields, having overlapping percentages not only on the boundaries (one grade step covers up to 42% but the next one begins at 40%), not covering all of the percentages (one ends at 38%, next begins at 40%), having the 'Min %' higher than the 'Max %'.
 c) Try all the abovementioned scenarios with mixed-up grade steps (the order from 0 to 100 is not kept as in the default grading scale for example). The order of the grade steps in the UI shouldn't matter as long as they are correctly formatted and cover all possible percentage values.
6. Try adding and deleting grade steps and performing step 5 again.
7. Try deleting the grading scale (the delete button should appear next to the save button only if a grading scale currently is saved in the database for the given course/exam)
8. Check if the requests passed to the server are correct, especially if the course/exam Ids are correctly set

### Test Coverage



grading-system.component.ts: 85%
grading-system.service.ts: 100%

### Screenshots


**Note:** Delete button appears only if a grading scale for the given course/exam exists

![Screenshot 2021-05-02 204327](https://user-images.githubusercontent.com/63168603/116822427-63e7c380-ab87-11eb-8748-04706cd254a7.png)


![Screenshot 2021-05-02 204401](https://user-images.githubusercontent.com/63168603/116822420-60543c80-ab87-11eb-89e4-33b2f8c861f2.png)


![Screenshot 2021-05-02 204418](https://user-images.githubusercontent.com/63168603/116822424-62b69680-ab87-11eb-9158-294bb603b737.png)


![Screenshot 2021-05-02 204431](https://user-images.githubusercontent.com/63168603/116822425-634f2d00-ab87-11eb-8c1d-ee70424d9210.png)

![Screenshot 2021-05-07 184518](https://user-images.githubusercontent.com/63168603/117479863-82740300-af69-11eb-9c51-6bee01547f96.png)

![Screenshot 2021-05-07 184553](https://user-images.githubusercontent.com/63168603/117479865-830c9980-af69-11eb-8d62-62ce44ea4f9f.png)
